File: house-robber/house-robber.cpp
class Solution {
public:
    int dp[105];
    int rob(vector<int>& nums) {
        for(int i = 0; i < 102; i++) {
            dp[i] = -1;
        }
        return solve(nums, 0);
    }
    
    int solve(vector<int>& nums, int pos)
    {
        if (pos >= nums.size()) {
            return 0;
        }
        
        if(dp[pos] != -1)
            return dp[pos];
        
     
        int q1 = solve(nums, pos + 1);
        int q2 = nums[pos] + solve(nums, pos + 2);
        
        return dp[pos] = max(q1, q2);
    }
};

File: house-robber/house-robber.py
class Solution:
    def __init__(self):
        self.dp = [-1 for i in range(110)]
        
    def rob(self, nums: List[int]) -> int:
        self.nums = nums
        return self.solve(0)
    
    def solve(self, pos) -> int:
        if pos >= len(self.nums):
            return 0
        
        if self.dp[pos] != -1:
            return self.dp[pos]
        
        self.dp[pos] = max(self.nums[pos] + self.solve(pos + 2), self.solve(pos + 1))
        return self.dp[pos]